### PR TITLE
ASoC: Intel: boards: use software node API

### PR DESCRIPTION
**Pull request by plbossart**
Updated from initial upstream contribution by Heikki Krogerius

### CODE REVIEW

**Comment by plbossart** [comment]
turns out the helper is not needed, as mentioned by @jwrdegoede in previous comments.

Rework and update (put_device() use was incorrect).

**Review by bardliao — APPROVED** [review]

**Review by lyakh — APPROVED** [review]
not really familiar with this API, but FWIW LGTM

**Comment by plbossart** [comment]
@jwrdegoede any comments? I struggled on the atom drivers since they all use a different structure, but I think this is correct now.

**Comment by plbossart** [comment]
also forgot to ask @andy-shev for feedback.

**Comment by andy-shev** [comment]
> also forgot to ask @andy-shev for feedback.

Code looks good. Just want to be sure I understand the lifetime constraints. You have a board file (where this code is applied) which lives shorter or longer than codec device. Is it correct? And actually which is the right one (shorter or longer)? In **some** cases one may use https://elixir.bootlin.com/linux/v5.12-rc5/source/drivers/base/swnode.c#L1076 but with care.

**Comment by plbossart** [comment]
> > also forgot to ask @andy-shev for feedback.
> 
> Code looks good. Just want to be sure I understand the lifetime constraints. You have a board file (where this code is applied) which lives shorter or longer than codec device. Is it correct? And actually which is the right one (shorter or longer)? In **some** cases one may use https://elixir.bootlin.com/linux/v5.12-rc5/source/drivers/base/swnode.c#L1076 but with care.

The platform devices and their drivers (modified here) depend on the functionality provided by the codec device/drivers, so the machine drivers will live for shorter time than the codec ones. The machine platform device is created AFTER we detect the presence of a codec device, and conversely we remove the machine platform device BEFORE removing the codec device. 

That was the main argument behind NOT using device managed APIs, as Heikki did in his initial patch submitted upstream, as we are adding the properties on the codec device but from the machine driver, so when we remove the machine device the properties are not removed. It's much better to add/remove properties in the same scope, we've been there before and using devm_ creates lots of issues in our load/unload tests. That's what we are doing for SoundWire machine drivers, but for old Baytrail stuff we didn't, so we have to add an explicit remove step in the machine platform driver.

**Comment by andy-shev** [comment]
> > > also forgot to ask @andy-shev for feedback.
> > 
> > 
> > Code looks good. Just want to be sure I understand the lifetime constraints. You have a board file (where this code is applied) which lives shorter or longer than codec device. Is it correct? And actually which is the right one (shorter or longer)? In **some** cases one may use https://elixir.bootlin.com/linux/v5.12-rc5/source/drivers/base/swnode.c#L1076 but with care.
> 
> The platform devices and their drivers (modified here) depend on the functionality provided by the codec device/drivers, so the machine drivers will live for shorter time than the codec ones. The machine platform device is created AFTER we detect the presence of a codec device, and conversely we remove the machine platform device BEFORE removing the codec device.
> 
> That was the main argument behind NOT using device managed APIs, as Heikki did in his initial patch submitted upstream, as we are adding the properties on the codec device but from the machine driver, so when we remove the machine device the properties are not removed. It's much better to add/remove properties in the same scope, we've been there before and using devm_ creates lots of issues in our load/unload tests. That's what we are doing for SoundWire machine drivers, but for old Baytrail stuff we didn't, so we have to add an explicit remove step in the machine platform driver.

Right, that explains how above patch is organazed and which APIs it uses. Thanks! To me it's clear and correct solution.

**Comment by plbossart** [comment]
Something's wrong here, there is a refcount error when removing the device software node

````
[  119.949838] bytcr_rt5640 bytcr_rt5640: before device remove software node
[  119.950358] ------------[ cut here ]------------
[  119.950395] refcount_t: underflow; use-after-free.
[  119.950483] WARNING: CPU: 1 PID: 1317 at lib/refcount.c:28 refcount_warn_saturate+0xa6/0xf0
[  119.950541] Modules linked in: snd_seq_dummy snd_hrtimer iptable_nat snd_soc_sst_bytcr_rt5640 intel_powerclamp iwlmvm iwlwifi mei_txe mei snd_sof_acpi_intel_byt(-) snd_sof_intel_ipc snd_sof_acpi snd_sof_xtensa_dsp snd_soc_rt5670 snd_sof snd_soc_rt5651 ledtrig_audio snd_soc_rt5645 snd_intel_sst_acpi snd_intel_sst_core snd_soc_rt5640 snd_soc_rl6231 snd_soc_sst_atom_hifi2_platform snd_soc_acpi_intel_match snd_soc_acpi regmap_i2c snd_soc_core snd_intel_dspcfg snd_intel_sdw_acpi snd_compress snd_seq snd_seq_device snd_pcm snd_timer snd soundcore zram mmc_block i915 i2c_algo_bit xhci_pci drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops xhci_hcd drm sdhci_acpi sdhci fuse
[  119.951352] CPU: 1 PID: 1317 Comm: rmmod Not tainted 5.12.0-rc4-test+ #8
[  119.951391] Hardware name: ZOTAC XXXXXX/Cherry Trail FFD, BIOS 5.11 09/28/2016
[  119.951419] RIP: 0010:refcount_warn_saturate+0xa6/0xf0
[  119.951465] Code: 05 0e 0a 3b 01 01 e8 41 92 65 00 0f 0b c3 80 3d fc 09 3b 01 00 75 95 48 c7 c7 80 17 5b 8f c6 05 ec 09 3b 01 01 e8 22 92 65 00 <0f> 0b c3 80 3d db 09 3b 01 00 0f 85 72 ff ff ff 48 c7 c7 d8 17 5b
[  119.951501] RSP: 0018:ffffb0c2813a3d78 EFLAGS: 00010286
[  119.951544] RAX: 0000000000000000 RBX: ffff9b80c5fd5e28 RCX: 0000000000000027
[  119.951638] RDX: 0000000000000027 RSI: 0000000058e38e39 RDI: ffff9b81349d7738
[  119.951676] RBP: ffff9b80c5936810 R08: ffff9b81349d7730 R09: 0000000000000000
[  119.951706] R10: 0000000000000001 R11: 0000000000000001 R12: 0000000000000000
[  119.951735] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[  119.951764] FS:  00007f38a6fe4b80(0000) GS:ffff9b8134800000(0000) knlGS:0000000000000000
[  119.951799] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  119.951829] CR2: 00007f38a7051b40 CR3: 000000003eb1a000 CR4: 00000000001006e0
[  119.951860] Call Trace:
[  119.951896]  snd_byt_rt5640_mc_remove+0x2f/0x43 [snd_soc_sst_bytcr_rt5640]
[  119.951963]  platform_remove+0x1a/0x40
[  119.952014]  device_release_driver_internal+0xf1/0x1d0
[  119.952065]  bus_remove_device+0xed/0x160
[  119.952113]  device_del+0x187/0x3e0
[  119.952169]  platform_device_del.part.0+0xe/0x60
[  119.952216]  platform_device_unregister+0x17/0x30
[  119.952263]  snd_sof_device_remove+0x53/0xf0 [snd_sof]
[  119.952367]  sof_acpi_remove+0x2f/0x40 [snd_sof_acpi]
[  119.952414]  platform_remove+0x1a/0x40
[  119.952459]  device_release_driver_internal+0xf1/0x1d0
[  119.952508]  driver_detach+0x42/0x90
[  119.952554]  bus_remove_driver+0x56/0xd0
[  119.952667]  __x64_sys_delete_module+0x18f/0x250
[  119.952728]  ? exit_to_user_mode_prepare+0x2f/0x130
[  119.952784]  do_syscall_64+0x33/0x40
[  119.952827]  entry_SYSCALL_64_after_hwframe+0x44/0xae
[  119.952869] RIP: 0033:0x7f38a711135b
[  119.952904] Code: 73 01 c3 48 8b 0d 15 1b 0c 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 90 f3 0f 1e fa b8 b0 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d e5 1a 0c 00 f7 d8 64 89 01 48
[  119.952939] RSP: 002b:00007ffc837cd388 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
[  119.952983] RAX: ffffffffffffffda RBX: 000055969ecf0760 RCX: 00007f38a711135b
[  119.953013] RDX: 000000000000000a RSI: 0000000000000800 RDI: 000055969ecf07c8
[  119.953042] RBP: 0000000000000000 R08: 0000000000000000 R09: 0000000000000000
[  119.953070] R10: 00007f38a7184ac0 R11: 0000000000000206 R12: 00007ffc837cd5c0
[  119.953098] R13: 00007ffc837cd89a R14: 000055969ecf02a0 R15: 000055969ecf0760
[  119.953178] ---[ end trace 3f804da710f70803 ]---
[  119.953213] bytcr_rt5640 bytcr_rt5640: before after remove software node
````

Not sure what's happening here. full dmesg here 
[dmesg.txt](https://github.com/thesofproject/linux/files/6299525/dmesg.txt)

**Comment by plbossart** [comment]
same issue with SoundWire devices, and this was already reported in CI results but the results are PASS
https://sof-ci.01.org/linuxpr/PR2810/build5522/devicetest/?model=TGLU_RVP_SDW&testcase=check-kmod-load-unload

@marc-hb @fredoh9 @aiChaoSONG any idea why we pass a test with such a blatant refcount issue?

**Comment by marc-hb** [comment]
> https://sof-ci.01.org/linuxpr/PR2810/build5522/devicetest/?model=TGLU_RVP_SDW&testcase=check-kmod-load-unload
> @marc-hb @fredoh9 @aiChaoSONG any idea why we pass a test with such a blatant refcount issue?

Probably because we don't have a culture of "testing the tests" and we're just not trying hard to find issues in general. A test that never reported any failure hasn't been tested and probably does not work, yet for some reasons we consider passing tests as good enough to make even major tests changes. I've been trying to change this "green failure" culture for some time but finally understood it's not anyone's priority. We didn't even have CI for sof-test when I joined, so not all is bad :-)

I see the dmesg.txt at that link does not have any "error" keyword, that could be why. What is the log level of this Call Trace? 

----

We love extending `ignore_str` to avoid looking at issues we want to believe don't matter but that seems to be unfortunately the only time we do some "test testing". There are extremely few people contributing to sof-test with code and reviews (just check the https://github.com/thesofproject/sof-test/graphs/contributors and other stats) it is not considered as a priority. The quality bar is also fairly low, for instance:

- When I added `fake_kern_error()` to "test the tests" in https://github.com/thesofproject/sof-test/pull/361 the first reaction was "why this fake_kern_error function is needed? And where we will use this function?"

- This old  journalctl refactoring attempt  didn't test the tests and failed to find kernel errors in `check-playback.sh`: https://github.com/thesofproject/sof-test/pull/354#pullrequestreview-481513488

- The huge `journalctl` refactoring in https://github.com/thesofproject/sof-test/pull/529 was auto-approved without review ("will merge blindly and fix later") because "the tests were passing": meaning they didn't find any error - which does not prove anything.

- A quote from the latest and current `journalctl` refactoring https://github.com/thesofproject/sof-test/pull/639#issuecomment-815399612
> Why are some other cases have DMESG_LOG_START_LINE? actually I don't know, ask the owner!

- Also unfinished https://github.com/thesofproject/sof-test/issues/312,  https://github.com/thesofproject/sof-test/issues/489

Etc., these are just some relevant examples from the top of my head.

**Comment by plbossart** [comment]
The first patch in this PR was merged by GregKH, so we can merge and contribute the two other patches upstream after v5.13-rc1

**Comment by plbossart** [comment]
rebased to make SoundWire error go away

**Comment by plbossart** [comment]
Rebased on top of v5.13-rc1 to make initial commit go away. This should be good to merge and send upstream now.

File: sound/soc/intel/boards/bytcr_rt5640.c [hunk 1]
```diff
@@ -960,7 +963,18 @@ static int byt_rt5640_add_codec_device_props(const char *i2c_dev_name)
 	if (byt_rt5640_quirk & BYT_RT5640_JD_NOT_INV)
 		props[cnt++] = PROPERTY_ENTRY_BOOL("realtek,jack-detect-not-inverted");
 
-	ret = device_add_properties(i2c_dev, props);
+	fwnode = fwnode_create_software_node(props, NULL);
+	if (IS_ERR(fwnode)) {
+		/* put_device() is not handled in caller */
+		put_device(i2c_dev);
```
Comment (lyakh) [inline sound/soc/intel/boards/bytcr_rt5640.c:969]: would it be desirable to unify this with bytcr_rt5651? If you decide to add that to this PR, you might as well remove the spurious empty line added at the end of `snd_byt_cht_es8316_mc_remove()` above

**Review by lyakh — COMMENTED** [review]
nothing significant, maybe a minor future clean up

File: sound/soc/intel/boards/bytcr_rt5640.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/bytcr_rt5640.c:969]: @lyakh the two machine drivers handle properties in different ways, I am not sure if they can be unified. That would be for @jwrdegoede to check, the code came from his contributions.

Will remove the spurious newline.

Comment (lyakh) [reply sound/soc/intel/boards/bytcr_rt5640.c:969]: Sorry, I didn't mean unifying the two drivers completely, I just meant the handling of `put_device()` - either within or outside of `byt_rt5651_add_codec_device_props()` and `byt_rt5640_add_codec_device_props()` respectively. But yes, a minor thing anyway

Comment (plbossart) [reply sound/soc/intel/boards/bytcr_rt5640.c:969]: The structure is rather different so I am not sure if we can unify. The functions have different prototypes so it's not small surgery I am afraid.

**Comment by plbossart** [comment]
I'll go ahead and merge, this PR has been around for quite some time and I want to send the patches upstream.

**Pull request merged**